File: 백준/Silver/2178. 미로 탐색/미로 탐색.py
import sys
input=sys.stdin.readline

def find():
    q.append([0,0,0])
    while len(q)>0:
        i,j,count=q.pop(0)
        if visited[i][j]>count+1:
            count=count+1
            visited[i][j]=count
            if i==N-1 and j==M-1:
                print(count)
                return # 함수 나가기
            if i+1<N and l[i+1][j]=='1':
                q.append([i+1,j,count])
            if j+1<M and l[i][j+1]=='1':
                q.append([i,j+1,count])
            if i-1>=0 and l[i-1][j]=='1':
                q.append([i-1,j,count])
            if j-1>=0 and l[i][j-1]=='1':
                q.append([i,j-1,count])

N,M=map(int,input().split())
l=[[]for i in range(N)]
for i in range(N):
    l[i]=list(input().rstrip())

q=[]
total=[]
visited=[[10001 for i in range(M)] for i in range(N)]
find()

File: 백준/Silver/2178. 미로 탐색/미로 탐색.js
const readline=require('readline')
const rl=readline.createInterface({
  input:process.stdin,
  output:process.stdout
})

const input=[]
let iter=0
rl.on('line',(line)=>{
  input.push(line)
}).on('close',()=>{
  const [N,M]=input[iter++].split(' ').map(Number)
  const l=[]
  for(let i=0;i<N;i++){
    l.push(input[iter++].split('').map(Number))
  }

  const arr=[]
  const maxValue=N*M
  for(let i=0;i<N;i++){
    arr.push([...new Array(M)].map(()=>maxValue))
  }
  
  const xbox=[0,0,-1,1]
  const ybox=[-1,1,0,0]

  arr[0][0]=1
  const q=[[0,0]]
  while(q.length>0){
    const [x,y]=q.shift()
    if (x===N-1 && y===M-1) {
      console.log(arr[x][y])
      process.exit()
    }
    const now=arr[x][y]
    for(let i=0;i<4;i++){
      const newx=x+xbox[i]
      const newy=y+ybox[i]
      if (0<=newx && newx<N && 0<=newy && newy<M && l[newx][newy]!==0){
        if (arr[newx][newy]===maxValue) {
          arr[newx][newy]=now+1
          q.push([newx,newy])
        }
      }
    }
  }

  process.exit()
})
